CMakeLists.txt

```diff
@@ -163,6 +163,48 @@ option(BUILD_TESTING "Whether to enable tests" ON)
 
 option(BUILD_W3CTRACECONTEXT_TEST "Whether to build w3c trace context" OFF)
 
+option(MAINTAINER_MODE "Build in maintainer mode (-Wall -Werror)" OFF)
+
+if(MAINTAINER_MODE)
+  add_compile_options(-Wall)
```

Comment:
We can add a section for windows, for developers.

Something like:
```
  if(CMAKE_CXX_COMPILER_ID MATCHES "MSVC")

    # Tested with MSVC ??.?? on github.
    if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL ??.??)
      # Relaxed warnings

      # Enforced warnings
    endif()

  endif()
```

@owent 

Could you provide the exact MSVC version to check for, and the warnings to enable ?

This will allow to do a build locally in MAINTAINER_MODE with MSVC.

Thanks.

Reply:
We have a toolset to use `/W4 /wd4100 /wd4125 /wd4566 /wd4127 /wd4512 /WX` for all codes(https://github.com/atframework/cmake-toolset/blob/bfe3a3249c9c32c7a9a75c9e078ba3e04acfe5d2/CompilerOption.cmake#L646).
And we disable more warnings for codes generated by protobuf.(e.g. `/wd4244 /wd4251 /wd4267 /wd4309 /wd4668 /wd4946 /wd4090`) and some warnings (`/wd5054 /wd4996 /4800`) are also disabled by MSVC_VERSION. (https://github.com/atframework/cmake-toolset/blob/bfe3a3249c9c32c7a9a75c9e078ba3e04acfe5d2/modules/ProjectBuildTools.cmake#L1032) 
I'm not sure if these are enough for grpc,but we use this toolset to build otel-cpp and all dependencies of it in our project.

Reply:
Thanks for the details.

The following is now in maintainer mode for msvc, in the main CMakeList.txt:
`/W4 /wd4100 /wd4125 /wd4566 /wd4127 /wd4512 /WX`

About additional warnings for grpc, they seem to be all there already except `/wd4090`, in file
`exporters/otlp/include/opentelemetry/exporters/otlp/protobuf_include_prefix.h`

Will add `/wd4090` if it gets reported by the build.